### PR TITLE
[Test]:Add FP8 quantization support and A/B gates for MammothModa2

### LINKED ISSUE #7125: [Performance]: Enable supported weight/activation quantization formats and validate generation experts, extra vocabulary/head, checkpoint loading, and quantization scales.
### Proposal to improve performance

Parent: #7075 — sub-issue for the AR-stage quantization TODO.

Enable FP8 (dynamic W8A8) and INT8 (W8A8) quantization for the MammothModa2 AR stage to reduce AR prefill/decode latency and peak GPU memory.

## Owner and scope

**Owner:** @wenj-yan

**In scope (MammothModa2 AR stage):**

- Enable supported weight/activation quantization formats for the AR LLM, starting with FP8 (dynamic W8A8) and INT8 (W8A8), via the unified `build_quant_config()` entry (#1763 / #1764).
- Validate **generation experts** (generation-specific expert/FFN paths are quantized correctly; routers/gates stay in full precision per the convention in #1854).
- Validate the **extra vocabulary / lm_head**: correct quantization exclusion and prefix handling for the extended vocab and generation head, so exclude-lists actually hit (precedent: lm_head prefix fix in #4528; unquantized lm_head GEMM issue in #1790).
- Validate **checkpoint loading** for pre-quantized checkpoints: weight/scale loading, packed-module mapping, and scale routing to the AR component only (precedent for wrong-scope routing: #2686; scale remapping: #5087).
- Validate **quantization scales**: scale dtype/shape sanity, NaN/clamp handling on load.
- Document the supported AR quantization configurations in the per-stage support matrix.

**Exclusions (owned elsewhere or out of scope):**

- DiT/transformer quantization — @xRay2016 (#7075 DiT quantization TODO).
- Quantized AR KV cache — @JiahengX.

## Plan

- **PR plan:**
  - PR 1 — Enable online FP8 (W8A8 dynamic) for the AR stage: quant_config plumbing, generation-expert handling, lm_head/extra-vocab exclusion, unit tests.
  - PR 2 — INT8 (W8A8) + pre-quantized checkpoint loading: scale loading/remapping, scale validation, loader tests.
  - PR 3 — Validation & docs: quality/parity test suite, support-matrix update.
- **PR ready for review:** 2026-09-14 (UTC)
- **Target merge:** 2026-09-17 (UTC), in line with the umbrella checkpoint.

## Correctness / quality validation

- Numerical-parity tests for quantized (FP8/INT8) vs BF16 AR logits on mixed-token routing and generation constraints.
- AR-stage output (logits / hidden states) alignment checks.
- Checkpoint-loading tests: pre-quantized INT8 checkpoints load without scale errors; quantized weights land only on AR modules.

## Known blockers / constraints

- Pre-quantized INT8/FP8 MammothModa2 checkpoints may need to be produced locally (no official quantized checkpoint known).
- MammothModa2 has two versions, Preview and Dev, with different AR-LLM architectures. Both must be validated independently as separate baselines.

## Reviewer

Requesting maintainer review.

Refs #7075

### Report of performance regression

N/A

## Hardware / baseline / performance validation plan

- **Hardware/backend:** NVIDIA CUDA, RTX PRO 6000 / 96 GB ×1.
- **Model version note:** MammothModa2 has two versions, Preview and Dev, with different AR-LLM architectures. Both must be validated independently as separate baselines.
- **Baseline:**
  - BF16 baseline from @quan-yi-ai, using AR logits/hidden-state outputs as reference.
  - Preview and Dev versions baselined separately; AR quantization validation covers both.
- **Performance evidence (AR-only scope):**
  - AR prefill/decode latency and TPS, peak GPU memory.
  - A/B comparison against the BF16 baseline from @quan-yi-ai.
  - Preview and Dev versions recorded separately.

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by wenj-yan**
Resolves #7125 (FP8 portion). Part of #7075.

## Purpose
Enable online FP8 (dynamic W8A8) quantization for the MammothModa2 AR stage,computed directly from the BF16 checkpoint — no pre-quantized weights required.This reduces AR prefill/decode latency and peak GPU memory, and corresponds to PR 1 of the plan in #7125.

Tests added in `test_mammoth_moda2_fp8_quantization.py` (GPU E2E A/B gates,BF16 vs FP8 under the same seed with greedy decoding):

- `test_ar_generation_smoke[bf16|fp8]` — each format loads and produces a non-empty greedy sequence;
- `test_bf16_vs_fp8_generation_consistency` — AR text A/B: token agreement  ≥ 0.90, top-1 logprob cosine ≥ 0.90;
- `test_bf16_vs_fp8_t2i_image_consistency` — t2i image A/B with FP8 applied to  the AR stage only (stage 0), DiT kept in BF16: decoded-image cosine ≥ 0.99,  rel-L2 < 0.05.

CPU-only structural quantization unit tests live in
`tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_quantization.py`.

Per #7125, both MammothModa2 versions (Preview and Dev, which have different AR-LLM architectures) are validated independently.
## Test Plan
```bash
pytest tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_fp8_quantization.py -v -s
pytest tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py -v -s
```
All cases use seed=42 and greedy decoding (temperature=0, top_k=1); t2i uses 256×256 images with 2 diffusion steps (CI-speed configuration). The only A/B variable is the stage-0 quantization key.
**vLLM Version:0.28.0

**vLLM-Omni Commit:g3204a0b2a (0.1.dev2892)

## Test Result

**Hardware:** NVIDIA RTX PRO 6000 96 GB ×1, Python 3.12.

### MammothModa2-Preview — 4 passed in 288s ✅

| Test | Metric | Threshold | Measured | Result |
| :--- | :--- | :--- | :--- | :--- |
| smoke [bf16] / [fp8] | load + generate | — | — | ✅ |
| AR text A/B | token agreement | ≥ 0.90 | 1.0000 (32/32 identical) | ✅ |
| | top-1 logprob cosine | ≥ 0.90 | 0.9968 | ✅ |
| | logprob MAE | observed | 0.0338 | — |
| t2i image A/B | image cosine | ≥ 0.99 | 0.999898 | ✅ |
| | rel-L2 | < 0.05 | 0.0153 | ✅ |
| | PSNR / MAE / max-abs / pixel-match | observed | 43.36 dB / 0.0048 / 0.035 / 68.8% | — |

### MammothModa2-Dev — 4 passed ✅

| Test | Metric | Threshold | Measured | Result |
| :--- | :--- | :--- | :--- | :--- |
| smoke [bf16] / [fp8] | load + generate | — | — | ✅ |
| AR text A/B | token agreement | ≥ 0.90 | 1.0000 (32/32 identical) | ✅ |
| | top-1 logprob cosine | ≥ 0.90 | 0.9447 | ✅ |
| | logprob MAE | observed | 0.0210 | — |
| t2i image A/B | image cosine | ≥ 0.99 | 0.999980 | ✅ |
| | rel-L2 | < 0.05 | 0.0115 | ✅ |
| | PSNR / MAE / max-abs / pixel-match | observed | 43.96 dB / 0.0052 / 0.055 / 59.2% | — |

### Conclusion

- On **Preview**, FP8 greedy token sequences are identical to BF16 (agreement 1.0), with logprob cosine 0.9968.
- On both **Preview** and **Dev**, decoded-image differences are sub-pixel (PSNR > 43 dB, cosine > 0.9998); even under the strict pixel_match definition (diff < 1/255, one uint8 step), 59–69% of pixels are exactly identical and the differences are imperceptible.
- The generation experts (`gen_mlp`) and the extra vocabulary/head (`gen_embed_tokens` / `gen_head`) behave correctly under FP8; all A/B gates pass on both model versions.

### Notes

- INT8 (W8A8) and pre-quantized checkpoint loading are left to PR 2 per the plan in #7125; DiT quantization is owned by @xRay2016 under #7075.

### CODE REVIEW

**Comment by lishunyang12** [comment]
is quantization already supported?

**Comment by wenj-yan** [comment]
Yes, the original quantization link can work normally.Both FP8 and INT8 tests are fine.

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+@pytest.mark.omni
+def test_bf16_vs_fp8_generation_consistency():
+    """A/B: BF16 vs FP8 — report token agreement + logprob similarity + MAE."""
+    bf16_ids, bf16_lp = _generate(MODEL_PATH, None)
+    fp8_ids, fp8_lp = _generate(MODEL_PATH, "fp8")
+
+    assert bf16_ids, "BF16 baseline produced no tokens"
+    assert fp8_ids, "FP8 run produced no tokens"
+
+    common = min(len(bf16_ids), len(fp8_ids))
```
Comment (JiahengX) [inline tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Could we also check the sequence lengths here? Right now this only compares the common prefix, so if BF16 generates 32 tokens and FP8 only generates 1 matching token, the agreement would still be 100%.

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    logprob_cos = _cosine_sim(bf16_lp[:lp_common], fp8_lp[:lp_common]) if lp_common > 0 else float("nan")
+    logprob_mae = _mean_abs_diff(bf16_lp[:lp_common], fp8_lp[:lp_common]) if lp_common > 0 else float("nan")
+
+    print(f"[FP8 A/B] token_agreement={token_agree:.4f} logprob_cosine={logprob_cos:.4f} logprob_mae={logprob_mae:.4f}")
+
+    assert token_agree >= MIN_TOKEN_AGREEMENT, (
+        f"BF16/FP8 greedy sequences diverge too much: agreement={token_agree:.3f} < "
+        f"{MIN_TOKEN_AGREEMENT}. bf16={bf16_ids} fp8={fp8_ids}"
+    )
+    if logprob_cos == logprob_cos:  # not NaN
```
Comment (JiahengX) [inline tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py:175]: Looks like the logprob check is skipped when `logprob_cos` is NaN. Would it make sense to assert that both sides actually have valid logprobs first, so the test can't pass without checking this metric?

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py (hunk 1 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Thank you for your check. Indeed, no length constraint was added. I have now modified the constraint conditions.Done in f4cb08c

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py (hunk 2 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py:184]: I re-checked and no NaN ever occurs here, and the new asserts produce identical results, so nothing was actually being skipped. Still, failing loudly is the right behavior, fixed in f4cb08c.The new interpretation results are [FP8 A/B] token_agreement=1.0000 logprob_cosine=0.9740 logprob_mae=0.0156

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+import pytest
+import torch
+
+# ---------------------------------------------------------------------------
+# Constants
+# ---------------------------------------------------------------------------
+
+# Local Dev checkpoint; override via env for CI / hub ids.
+MODEL_PATH = os.environ.get("MAMMOTH_MODA2_MODEL", "/root/autodl-fs/MammothModa2-Dev")
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Parametrize Preview and Dev hub IDs so both variants run outside your local setup.

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+
+# Minimum cosine similarity between BF16 and quantized top-1 logprob sequences.
+# FP8 W8A8 typically lands ~0.94-0.99 on greedy logprobs (measured 0.9447 on
+# RTX PRO 6000); keep a margin below that to catch real breakage, not normal
+# quantization drift.
+MIN_LOGPROB_COSINE = 0.90
+
+_PROMPT = "Explain multimodal generation in three sentences."
+
+pytestmark = [pytest.mark.slow]
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Add hardware markers and register these A/B tests in the GPU CI pipeline.

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py (hunk 3 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Thank you for your review! The standard checkpoint path has been modified in 3b89779.

File: tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py (hunk 4 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_fp8_quantization.py]: Thank you for your review!I have register the teset in the CI pipeiline in 27e4178
